playerctl/playerctl-cli.c

```diff
@@ -16,279 +16,359 @@
-  { NULL }
-};
+#define LENGTH(array) (sizeof array / sizeof array[0])
 
-/* returns a newline delimitted list of player names */
-static gchar *list_player_names(GError **err);
+G_DEFINE_QUARK(playerctl-cli-error-quark, playerctl_cli_error);
 
-int main (int argc, char *argv[])
+/* The player being controlled. */
+static gchar *player_name = NULL;
+/* The commands passed on the command line, filled in via G_OPTION_REMAINING. */
+static gchar **command = NULL;
+
+static gchar *list_player_names(GError **err)
 {
-  // seems to be required to print unicode (see #8)
-  setlocale(LC_CTYPE, "");
+  GError *tmp_error = NULL;
 
-  GOptionContext *context = NULL;
-  GError *error = NULL;
-  gboolean succes;
+  GDBusProxy *proxy = g_dbus_proxy_new_for_bus_sync(
+      G_BUS_TYPE_SESSION,
+      G_DBUS_PROXY_FLAGS_NONE,
+      NULL,
+      "org.freedesktop.DBus",
+      "/org/freedesktop/DBus",
+      "org.freedesktop.DBus",
+      NULL,
+      &tmp_error);
 
-  context = g_option_context_new("- Controller for MPRIS players");
-  g_option_context_add_main_entries(context, entries, NULL);
-  g_option_context_set_description(context, description);
-  g_option_context_set_summary(context, summary);
+  if (tmp_error != NULL) {
+    g_propagate_error(err, tmp_error);
+    return NULL;
+  }
 
-  succes = g_option_context_parse(context, &argc, &argv, &error);
-  g_option_context_free (context);
+  GVariant *reply = g_dbus_proxy_call_sync(proxy,
+      "ListNames",
+      NULL,
+      G_DBUS_CALL_FLAGS_NONE,
+      -1,
+      NULL,
+      &tmp_error);
 
-  if (!succes) {
-    g_printerr("Option parsing failed: %s\n", error->message);
-    return 1;
+  if (tmp_error != NULL) {
+    g_propagate_error(err, tmp_error);
+    g_object_unref(proxy);
+    return NULL;
   }
 
-  if (version_opt) {
-    g_print("v%s\n", PLAYERCTL_VERSION_S);
-    return 0;
+  GString *names_str = g_string_new("");
+  GVariant *reply_child = g_variant_get_child_value(reply, 0);
+  gsize reply_count;
+  const gchar **names = g_variant_get_strv(reply_child, &reply_count);
+
+  for (int i = 0; i < reply_count; i += 1) {
+    if (g_str_has_prefix(names[i], "org.mpris.MediaPlayer2")) {
+      g_string_append_printf(names_str, "%s\n", names[i] + 23);
+    }
   }
 
-  if (list_all_opt) {
-    gchar *player_names = list_player_names(&error);
+  g_object_unref(proxy);
+  g_variant_unref(reply);
+  g_variant_unref(reply_child);
+  g_free(names);
 
-    if (error != NULL) {
-      g_printerr("Could not list players: %s\n", error->message);
-      g_clear_error(&error);
-      return 1;
-    }
+  return g_string_free(names_str, FALSE);
+}
 
-    if (player_names[0] == '\0')
-      g_printerr("%s\n", "No players were found");
-    else
-      g_print("%s", player_names);
-    g_free(player_names);
+static gboolean list_all_players_and_exit (const gchar *name, const gchar *value, PlayerctlPlayer *player, GError **error)
```

Comment:
I agree, hence I named it `_and_exit`. We are bound to this function signature because of [`GOptionArgFunc`](https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html#GOptionArgFunc). I agree it's wonky, but we're not the only one doing it like this, for example there's [Eye of GNOME](https://git.gnome.org/browse/eog/tree/src/main.c#n50). Would it be better to add a compiler annotation [`noreturn`](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html)?